src/controlplane/scraper.go

```diff
@@ -0,0 +1,282 @@
+		return nil, fmt.Errorf("fetching K8s version: %w", err)
+	}
+
+	// Building pod lister and closers for pod autodisover.
+	s.buildLister()
+
+	return s, nil
+}
+
+// Run scraper collect the data populating the integration entities.
+func (s *Scraper) Run(i *integration.Integration) error {
+	var jobs []*scrape.Job
+
+	for _, component := range s.components {
+		var job *scrape.Job
+
+		var err error
+
+		// Static endpoint take precedence over autodisover and fails if external endpoint
+		// cannot be scraped.
+		if component.StaticEndpointConfig != nil {
+			s.logger.Debugf("Building %q external endpoint job")
+
+			job, err = s.externalEndpoint(component)
+			if err != nil {
+				return fmt.Errorf("configuring %q external endpoint: %w", component.Name, err)
+			}
+		} else {
+			s.logger.Debugf("Building %q autodiscovered endpoint job")
+
+			job, err = s.autodiscover(component)
+			if err != nil {
+				return fmt.Errorf("autodiscovering %q endpoint: %w", component.Name, err)
+			}
+		}
+
+		// If autodisover do not find any valid endpoint it will return a nil job and no error.
+		if job != nil {
+			jobs = append(jobs, job)
+		}
+	}
+
+	for _, job := range jobs {
+		s.logger.Debugf("Running job: %s", job.Name)
+
+		result := job.Populate(i, s.config.ClusterName, s.logger, s.k8sVersion)
+
+		if len(result.Errors) > 0 {
+			s.logger.Infof("Error populating data from %s: %v", job.Name, result.Error())
+		}
+	}
+
+	return nil
+}
+
+// externalEndpoint builds the client based on the StaticEndpointConfig and fails if
+// the client probe cannot reach the endpoint.
+func (s *Scraper) externalEndpoint(c component) (*scrape.Job, error) {
+	connector, err := controlplaneClient.DefaultConnector(
+		[]config.Endpoint{*c.StaticEndpointConfig},
+		s.K8s,
+		s.inClusterConfig,
+		s.logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("control plane component %q failed creating connector: %v", c.Name, err)
+	}
+
+	client, err := controlplaneClient.New(connector, controlplaneClient.WithLogger(s.logger))
+	if err != nil {
+		return nil, fmt.Errorf("creating client for component %s failed: %v", c.Name, err)
+	}
+
+	u, err := url.Parse(c.StaticEndpointConfig.URL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing static endpoint url for component %s failed: %v", c.Name, err)
+	}
+
+	grouper := grouper.New(
+		client,
+		c.Queries,
+		s.logger,
+		u.Host,
+	)
+
+	return scrape.NewScrapeJob(string(c.Name), grouper, c.Specs), nil
+}
+
+// autodiscover will iterate over the Autodiscovery configs from a component and for each:
+//  - Discover if any pod matches the selector.
+//  - Build the client, which probes all the endpoints in the list.
+// It uses the first autodiscovery config that can satisfy conditions above.
+// It doesn't fail if no autodiscovery satisfy the contitions.
+func (s *Scraper) autodiscover(c component) (*scrape.Job, error) {
+	for _, autodiscover := range c.AutodiscoverConfigs {
+		podName, err := s.discoverPod(autodiscover)
```

Comment:
I would make the discovery operation return a `*corev1.Pod`. Right now we're only using the labels but in the future we might want to check more info.